CanonicalABI.md

```diff
@@ -0,0 +1,1166 @@
+      gen.requeue(val) # put `val` back in the generator to get yielded first in `lower`
+      [val_ptr, val_len] = lower(direction, dst, val)
+      write_4_bytes(dst, ptr, val_ptr)
+      write_4_bytes(dst, ptr + 4, val_len)
+    }
+
+    flags<$fields>(val) => {
+      gen.requeue(val)
+      for i, bits in enumerate(lower(direction, dst, val)):
+        write_4_bytes(dst, ptr + i * 4, bits)
+    }
+
+    record_start<$fields> => {
+      for $field in $fields:
+        ptr = align_to(ptr, align(direction, $field))
+        lower_to_memory(direction, dst, ptr, gen)
+        ptr += size(direction, $field)
+      assert_eq(gen.next(), record_end<$fields>) # true by construction
+    }
+
+    variant_start<$cases>(discrim) => {
+      match variant_discriminant($cases) {
+        i8 => {
+          write_1_byte(dst, ptr, discrim)
+          ptr += 1
+        }
+        i16 => {
+          write_2_bytes(dst, ptr, discrim)
+          ptr += 2
+        }
+        i32 => {
+          write_4_bytes(dst, ptr, discrim)
+          ptr += 4
+        }
+        i64 => {
+          write_8_bytes(dst, ptr, discrim)
+          ptr += 8
+        }
+      }
+
+      # Recursively write the payload, if present for this case. Note that
+      # we align the pointer to the whole variant's alignment to ensure that
+      # all payloads start at the same address.
+      if $cases[discrim].has_payload:
+        ptr = align_to(ptr, align(direction, variant $cases))
+        lower_to_memory(direction, dst, ptr, gen)
+
+      assert_eq(gen.next(), variant_end<$cases>) # true by construction
+    }
+
+    handle<$resource>(val) => {
+      gen.requeue(val)
+      bits = lower(direction, dst, val)[0]
+      write_4_bytes(dst, ptr, bits)
+    }
+
+    push-buffer<$ty>(val) | pull-buffer<$ty>(val) => {
+      gen.requeue(val)
+      bits = lower(direction, dst, val)[0]
+      write_4_bytes(dst, ptr, bits)
+    }
+```
+
+## Helper `TABLE` class
+
+In Python-like pseudo-code this is how the `TABLE` referenced above in
+lifting/lowering will work for managing indices. Note that the implementation
+provided here is intended to be canonical. Specifically these properties are
+part of the canonical ABI:
+
+* There is a separate index space per-instance and per-resource type. This means
+  that each instance's adaptation effectively has access to a set of tables for
+  each hadnle type, and indices are managed by allocating from these tables.
+
+* Indices to refer to resources are handled in a LIFO fashion. Indexes are
+  allocated starting from zero and increasing afterwards as more space is
+  required. When an index is removed (because an instance said it no longer
+  needs a resource, then that index is queued up as the next to be allocated).
+
+* There can be at most 2^31-1 values of any handle type for a module. This is
+  done so indices can fit into a hypothetical `i31ref` in the future.
+
+```
+class TABLE:
+  instances = Map<(Instance, Resource), Slab<Handle>>
```

Comment:
It'd probably be good to implement Slab.  In particular, somewhere needs to clearly specify returning non-zero indices (as `insert_buffer` correctly does) since `push(handle)` is a bit unclear on this.